Design refresh

randyaskin opened the pull request:
This PR provides a rebuild of the px-dropdown component using iron-dropdown and iron-selector. It simplifies the API by removing the px-dropdown-content subcomponent and providing all of the configuration properties on the px-dropdown component itself. It is a rather large departure from the previous API, and will have to be updated and tested in px-demo, px-sass-doc, px-data-table, px-inbox and any other components that currently use px-dropdown. Also needs to be updated with the new buttons and form input designs as part of the current design refresh activities and signed off on by the designers.

Addresses issues #34 and #40.

mdwragg commented on sass/px-dropdown.scss:
```diff
@@ -0,0 +1,99 @@
+  border-color: var(--px-validation-success-text-color, $status-green3);
+  color: inherit;
+}
+.dropdown-trigger {
```
Suggest we BEM the names up here `.dropdown__trigger`?

mdwragg commented on sass/px-dropdown.scss:
```diff
@@ -0,0 +1,99 @@
+  margin-bottom: calculateRem(2px);
+  margin-left: $inuit-base-spacing-unit--tiny;
+}
+.dropdown-content {
```
Moar BEM.

mdwragg commented on sass/px-dropdown.scss:
```diff
@@ -0,0 +1,99 @@
+  max-width: var(--px-dropdown-max-width, 400px);
+  word-break: break-word;
+}
+.dropdown-option {
```
And here too (BEM)

mdwragg commented on px-dropdown.html:
```diff
@@ -129,478 +127,537 @@
-        this._reset();
-      }
+    _getElement: function(target) {
+      return document.querySelector(target);
```
Should drive this through the polymer API to make sure it work in shady/shadow dom.

mdwragg commented on px-dropdown.html:
```diff
@@ -129,478 +127,537 @@
+        var length = items.length,
+            computedItems = [],
+            i;
+        for(i=0; i<length; i++) {
```
Feel free to use the more semantically safe/easy to understand .forEach et al interators here, I doubt we're talking about large numbers of items here to perf is probably not a huge worry.

mdwragg commented on px-dropdown.html:
```diff
@@ -129,478 +127,537 @@
+    _setFocusedOption: function(newOption, oldOption) {
+      this._focusedOption = newOption;
+      if(newOption) {
+        newOption.classList.add('focused');
```
Prefer the use of `this.toggleClass()` rather than manually manipulating the classList .

mdwragg commented on px-dropdown.html:
```diff
@@ -129,478 +127,537 @@
-            break;
-        }
+      this._keyboardBeingUsed = true;
+      var keyPressed = event.detail.combo,
```
Make sure events are normalised to work in shadow/shady DOM.

mdwragg commented on px-dropdown.html:
```diff
@@ -129,478 +127,537 @@
+          // If something is focused, move to the previous sibling
+          if(focused > 0) {
+            this._setFocusedOption(options[focused - 1], options[focused]);
+            this.$.dropdown.querySelector('.dropdown-content').scrollTop -= options[focused].offsetHeight;
```
Drive through Polymer dom API to make sure safe across shadow/shady dom.

mdwragg commented on px-dropdown.html:
```diff
@@ -129,478 +127,537 @@
+      if (!this._keyboardBeingUsed) {
+        var currHighlightedItem = this.querySelector('.dropdown-option.focused');
+        if (currHighlightedItem) {
+          currHighlightedItem.classList.remove('focused');
```
As above, prefer `this.toggleClass`.

mdwragg commented on sass/px-dropdown.scss:
```diff
@@ -0,0 +1,99 @@
+  align-items: center;
+  justify-content: space-between;
+}
+.dropdown-trigger .label {
```
As above, more BEM loveliness.

mdwragg commented on sass/px-dropdown.scss:
```diff
@@ -0,0 +1,99 @@
+  text-overflow: ellipsis;
+  text-align: left;
+}
+.dropdown-trigger .icon {
```
As above, more BEM loveliness.

randyaskin commented:
To summarize our Slack conversation:

- BEM isn't possible, since iron-dropdown is expecting `.dropdown-trigger` and `.dropdown-content`
- iron-a11y-keys seems to take care of normalization with the `target` property, but I've updated the others
- replaced `for` loop with `forEach` and `classList` manipulation with `toggleClass` as requested
